packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -0,0 +1,201 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'constants.dart';
+
+/// Coordinates tab selection between a [TabBar] and a [TabBarView].
+///
+/// The [index] property is the index of the selected tab and the [animation]
+/// represents the current scroll positions of the tab bar and the tar bar view.
+/// The selected tab's index can be changed with [animateTo].
+///
+/// See also:
+///
+/// * [DefaultTabController], which simplifies sharing a TabController with
+/// its [TabBar] and a [TabBarView] descendants.
+class TabController extends ChangeNotifier {
+  /// Creates an object that manages the state required by [TabBar] and a [TabBarView].
+  TabController({ int initialIndex: 0, @required this.length, @required TickerProvider vsync })
+    : _index = initialIndex,
+      _previousIndex = initialIndex,
+      _animationController = new AnimationController(
+        value: initialIndex.toDouble(),
+        upperBound: (length - 1).toDouble(),
+        vsync: vsync
+   ) {
+    assert(length != null && length > 1);
+    assert(initialIndex != null && initialIndex >= 0 && initialIndex < length);
+  }
+
+  /// An animation whose value represents the current position of the [TabBar]'s
+  /// selected tab indicator as well as the scrollOffsets of the [TabBar]
+  /// and [TabBarView].
+  ///
+  /// The animation's value ranges from 0.0 to [length] - 1.0. After the
+  /// selected tab is changed, the animation's value equals [index]. The
+  /// animation's value can be [offset] by +/- 1.0 to reflect [TabBarView]
+  /// drag scrolling.
+  final AnimationController _animationController;
+  Animation<double> get animation => _animationController.view;
+
+  /// The total number of tabs. Must be greater than one.
+  final int length;
+
+  void _changeIndex(int value, { Duration duration, Curve curve }) {
+    assert(value != null);
+    assert(value >= 0 && value < length);
+    assert(duration == null ? curve == null : true);
+    if (value == _index)
+      return;
+    _indexIsChanging = true;
+    _previousIndex = index;
+    _index = value;
+    if (duration != null) {
+      _animationController
+        ..animateTo(_index.toDouble(), duration: duration, curve: curve).then((_) {
+          _indexIsChanging = false;
```

Comment:
I'm worried that someone could spam the index setter and get multiple index changes in flight and corrupt the state of `_indexIsChanging`.  Maybe that's not possible for some reason?  I'd probably make `_indexIsChanging` into an `int` and `++` it when we start a change and `--` it when the change is over.  We can keep the public API the same by testing it against zero.

Here's an example scenario that could get us in trouble.

1. Listen to the animation.
2. Set the index to a new value with zero duration.
3. In the animation listener callback, animate the index to a new value.

When we unwind from the animation listener callback, we'll set `_indexIsChanging` to false on line 66 even though the animation we kicked off in step 3 is still running.